PR #43 (add/getCustomFieldsValues) | integration tests only

rgadon107 opened the pull request:
## PR Summary

`get_custom_fields_values()` is a helper function called from `render_meta_box()` within the meta box generator of the Central Hub plugin. The latter function runs the business logic to populate the custom fields (post meta box) view in the WP admin. 

`get_custom_fields_values()` returns the filter `filter_meta_box_custom_fields` and 3 arguments (`$custom_fields`, `$meta_box_id`, and `$post_id` ). The filter allows the custom field values stored in the database to be modified before they are rendered to the metabox on the admin screen. 

The file path to `get_custom_fields_values()` (relative to the root of it's Git repo) is: 

>`wp-content/mu-plugins/central-hub/src/meta-data/helpers.php`.

This PR contains the _integration_ tests with WordPress for `get_custom_fields_values()`.

rgadon107 committed:
[WIP] Commit first & second integration tests.

rgadon107 committed:
First integration test: 

1) Edit test description.

2) Edit name of test method.

rgadon107 committed:
Edit second integration test:

1) Edit description of test method. 

2) Edit name of test method.

3) Modify the initial state of $config array.

4) Don't load $config into the config-store.

rgadon107 committed:
Edit second integration test:

1) Update $expected_custom_fields. 

2) Pass $expected_custom_fields as expectation to assertion.

rgadon107 committed:
Update use function declarations.

rgadon107 committed:
Add tearDown() method.

rgadon107 committed:
Edit third integration test.

1) Update description of test method.

rgadon107 committed:
- Clean up db following 2nd integration test. 

- Construct third integration test.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,151 @@
+	/**
+	 * Clean up the test environment after each test.
+	 */
+	public function tearDown() {
```
@rgadon107 This method isn't doing anything unique. Therefore, you can delete it.

rgadon107 replied:
Done.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,151 @@
+		add_post_meta( $this->post, 'event-time', '19:30:00' );
+		add_post_meta( $this->post, 'venue-name', 'Carnegie Hall' );
+
+		// Check database that post meta was added.
```
@rgadon107 Good job here!

rgadon107 replied:
Thank you!

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,151 @@
+			'venue-name' => 'Carnegie Hall',
+		];
+
+		$this->assertSame( $expected_custom_fields, get_custom_fields_values( $this->post, 'events', $config ) );
```
@rgadon107 Good job!

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,151 @@
+		$this->assertSame( $expected_custom_fields, get_custom_fields_values( $this->post, 'events', $config ) );
+
+		// Clean up database.
+		delete_post_meta( $this->post, 'event-date', '10-04-2019' );
```
@rgadon107 When deleting post meta, you don't have to pass in the value, ie 3rd argument.

rgadon107 replied:
@hellofromtonya Done.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,151 @@
+			'venue-name' => '',
+		];
+
+		$this->assertSame( $expected_custom_fields, get_custom_fields_values( $this->post, 'events', $config ) );
```
@rgadon107 Good job!

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,151 @@
+		$this->assertSame( $expected_custom_fields, get_custom_fields_values( $this->post, 'events', $config ) );
+	}
+}
+
```
@rgadon107 What about the `'filter_meta_box_custom_fields '` filter? You'll want to test that:

1. it fires
2. the expected arguments are received.

rgadon107 replied:
@hellofromtonya I can prove item 1 above. Not sure how to execute 2 above. 

I know that PHP provides a function `func_get_args()` that returns an array comprising a function's argument list. I've never properly used it before. Could it be applied in this situation?

hellofromtonya replied:
@rgadon107 Let me explain:

```php
$this->assertEquals( 20, has_filter( 'filter_meta_box_custom_fields' , function () {}, 20, 3 ); // returned false.
```

Let's go back to what [`has_filter()`](https://knowthecode.io/docx/wordpress/has_filter). This function requires the following:

- The name of the filter event
- The name of the function/method callback that is being checked, ie Is this callback registered to this filter event?
- Priority level

Notice that the 2nd argument is the exact callback that's registered to the filter event. Hold this in your mind.

A closure (which is also known as anonymous or unnamed function) does not have a user-defined name that is bound to the closure. Right? Why is this important?

Think about how PHP is managing memory. The callables (which includes functions, methods, and closures) are stored in memory. Named (ie the names we give to each callable) callables is stored in the Symbols table in memory and then bound to the memory location where the callable is stored. 

What about unnamed callables, ie closures? There's no name in the Symbols table. That means you can't reference it or call it by a name. Why? It doesn't have one. This distinction is important when you want to do something with that closure in relationship to a filter or action event.

What do I mean? `has_filter()` requires the name or memory location of the callback in order to _look up_ if that callback is registered (hooked) to the filter event. A closure doesn't have a name. It does have a memory location. Huh? Everything in memory has a specific location, which is used by PHP to point to that specific thing in memory. Notice in the Kint printout above that the closure's reference is not a name but instead a long hex code. That's the memory location/ID for that specific unnamed callback (closure).

Now back to your code:

```php
$this->assertEquals( 20, has_filter( 'filter_meta_box_custom_fields' , function () {}, 20, 3 ); // returned false.
```

You have a brand new closure as the 2nd argument. `function() {}` defines a brand new closure. It does _not_ point to the closure that is actually registered. 

## How can you reference a closure?

You have to get the memory ID/location. There are a couple of ways:

1. Storing the closure in a variable.
2. _Discover_ the memory ID/location by looking inside of `$wp_filter` to find the hex code. Messy.

Both of these approaches are messy and unnecessary. If you need to reference the callback, then use a named function instead.

## When to use a closure as a callback

As a general rule, you can use a closure as a callback is when that callback is:

1. Not reusable (not used anywhere else in the code, ie a true callback that is bound to one singular hook)
2. Will not be directly unhooked/unregistered, ie meaning it remains registered/hooked.

Therefore, in general, we say that a closure callback is not usable within `remove_filter()`, `has_filter()`, `remove_action()`, or `has_action()`.

hellofromtonya reviewed (changes requested):
Good job on the integration tests. There are few notes and a missing test.

This function also needs unit tests.

rgadon107 committed:
Per CR, remove tearDown() method. Not useful.

rgadon107 committed:
Remove optional $meta_value arguments.

rgadon107 committed:
Add 4th integration test: 

- `test_should_return_true_when_filter_tag_has_registered_callback`.

rgadon107 committed:
Update 4th integration test. 

- Register anonymous function to filter event. 
- Test expected callback args with func_get_args.
- Update custom fields and test with func_get_args again. 
- Check if any cb is registered to filter.
- Clean up by removing cb registered to filter.

rgadon107 committed:
Update .gitignore file.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,183 @@
+
+			$this->assertSame( $update_custom_fields, func_get_args() );
+		}, 20, 3 );
+
```
@rgadon107 Nothing in the closure will run. Why? The function `get_custom_fields_values()` does not run in this test.

How to resolve? You need to set up the test just as you've done in previous tests above:

- Define the `$config` for the custom fields
- Add post meta into the database, ie the values you will be tested against in the closure.
- Then run the function

```php
$this->assertSame( $updated_custom_fields, get_custom_fields_values( $this->post, 'events', $config ) );
```

Notice that you are testing that the updated custom fields do get returned. Why?

rgadon107 replied:
@hellofromtonya 
>Notice that you are testing that the updated custom fields do get returned. Why?

Because I'm filtering the value of `$custom_fields` in the array of arguments, and I want to ensure that those filtered values were returned to the calling function.

rgadon107 replied:
@hellofromtonya Thank you for the helpful hint here. I incorporated that strategy into my update of the 4th integration test.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,183 @@
+			];
+
+			$this->assertSame( $update_custom_fields, func_get_args() );
+		}, 20, 3 );
```
@rgadon107 This is a filter. Therefore, you need to return the updated custom fields as the last line in the closure. Why?

rgadon107 replied:
@hellofromtonya 
>Therefore, you need to return the updated custom fields as the last line in the closure. Why?

If I fail to return the updated values, then the filtered values will not get returned to the callback registered to the filter. Now I understand what's going on.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,183 @@
+		$this->assertTrue( has_filter( 'filter_meta_box_custom_fields' ) );
+
+		// Clean up.
+		$this->assertTrue( remove_all_filters( 'filter_meta_box_custom_fields', 20 ) );
```
@rgadon107 There's no need to assert true here. Why? `remove_all_filters` is _not_ an expected behavior within the `get_custom_fields_values()` function. Rather, you are using it to unregister (unhook) the callback from the filter that's within this function.

Does this make sense?

rgadon107 replied:
@hellofromtonya

>Does this make sense?

Yes. I thought I needed to `assertTrue` to confirm that the filter hook actually was removed. I'll remove it since it's not needed.

rgadon107 committed:
Replace anonymous (unnamed) function with named callback.

rgadon107 committed:
Add assertion to check priority of `has_filter()`.

rgadon107 commented:
@hellofromtonya Outstanding explanation above ^^^ with regard to the use of named custom callbacks versus closures when testing filters and actions, and _when_ to use which. I modified the 4th integration test accordingly and used a named custom callback instead of a closure. Wasn't sure initially if I could do that when I started to write that test. Now I know I can.

rgadon107 committed:
Update 4th integration test. 

1) Rename test method to refer to registered custom callback. 

2) Add initial custom fields $config.

3) Add post meta to database. 

4) Check that post meta was saved to database.

5) Register custom callback (named function) to filter event. 

6) Return filtered $custom_fields to registered callback. 

7) Pass filtered $custom_fields array values to variable. 

8) Run assertion to check that filtered cf values are returned with function under test is run. 

9) Check `has_filter()` for filter event and registered callback. 

10) Remove filter and delete post meta.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,220 @@
+		 * Filter the custom field values with a custom callback
+		 */
+		function filter_custom_field_values( $custom_fields ) {
+
```
@rgadon107 You want to test the following within the callback:

1. The callback receives only 3 arguments.
2. The values for each is as expected.

For assertion 1, use `func_get_args`. Why? You want to grab _all arguments passed to the callback_ to ensure only 3 are received.

Doesn’t defining the function params do this? No. Why? A function can receive _more_ arguments than it specifies in its params list (but it can’t receive less). Look at your code above. You’ve registered that the callback will receive 3 args, but you only defined 1 param, named `$custom_fields`. Although you only defined 1 params for the function, it will actually (should) receive 3.

Does that make sense?

rgadon107 replied:
@hellofromtonya 
>Does that make sense?

Yes. I updated the custom callback to pass in the 3 callback parameters. The filter passes in each one of them to the callback when the number of args is set to 3. I ran `func_get_args()` inside the cb and used `var_dump()` to check the value. The initial (unfiltered) values returned to `get_custom_field_values()` were all there. 

>For assertion 1, use func_get_args. Why? You want to grab all arguments passed to the callback to ensure only 3 are received.

I'm not sure how to write a valid assertion to test that the 3 unfiltered args were passed to the cb. There's a couple of reasons why this is problematic. 

1) PHPUnit will not allow me to run `$this` inside a non-object context (e.g. the custom callback registered to the filter). 

2) If I return `func_get_args()` to the custom cb, and then invoke the cb as an expected value within the assertion, PHPUnit throws a fatal error. The custom cb is also not a callable; I can't invoke the function name in a string as an expected value and get the assertion to run. 

=============================================

## Doesn't work

	function filter_custom_field_values( $custom_fields, $meta_box_id, $post_id ) {
			$args = func_get_args();
                        
                        $this->assertSame( $args, get_custom_field_values( $this->post, 'events', $config )
		}

----------------------------------------------------------------

	function filter_custom_field_values( $custom_fields, $meta_box_id, $post_id ) {
			return func_get_args();
		}

 $this->assertSame( filter_custom_field_values( $custom_fields, $meta_box_id, $post_id ),
			get_custom_fields_values( $post_id, 'events', $config ) );

----------------------------------------------------------------

	function filter_custom_field_values( $custom_fields, $meta_box_id, $post_id ) {
			return func_get_args();
		}

 $this->assertSame( 'filter_custom_field_values',
			get_custom_fields_values( $post_id, 'events', $config ) );

hellofromtonya replied:
@rgadon107 Why can’t the embedded function access `$this`? It is not a function of the object. Therefore, it does not have access to the object.

As I noted elsewhere, I’m not a fan of functions within a function (ie embedded functions). Why? Readability. Difficulty understanding scope, ie what is available to the embedded function.

I suggest moving the embedded function/callback to be a public method of the test class. Then `$this` is available. Why? By moving the function to a method, the callback is a method of the test object.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,220 @@
+	/**
+	 * Test get_custom_fields_values() should register a custom callback and return true with callback priority level.
+	 */
+	public function test_filter_registers_custom_callback_and_returns_true_with_cb_priority_level() {
```
@rgadon107 The name of this test is misleading. What is being tested here? You’re testing that the filter works as expected, meaning:

1. Args are passed to the callback and are expected, ie callback receives only the args it expects and each is correct.
2. The changed custom fields are passed back.

In essence, you are testing that the filter works as expected.

hellofromtonya commented on mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php:
```diff
@@ -0,0 +1,220 @@
+		$this->assertSame( 'Carnegie Hall', get_post_meta( $this->post, 'venue-name', true ) );
+
+		// Register custom callback to filter $tag with priority of 20 and 3 available arguments.
+		add_filter( 'filter_meta_box_custom_fields',  __NAMESPACE__ . '\filter_custom_field_values', 20, 3 );
```
@rgadon107 I’m not a fan of embedded functions. Instead, move this function to a method. Make the method public. 

How do you register a method to a filter event? 

```php
add_filter( ‘filter_meta_box_custom_fields’, [ $this, ‘filter_custom_field_values’ ] );
```

hellofromtonya committed:
Updated testing filter callback.

1. Changed the embedded function to be a public callback on the test object.
2. Changed the registration for the new method.
3. Within the callback,
   a. tested the expected arguments
   b. modified the custom fields values.

hellofromtonya committed:
Changed test method.

hellofromtonya committed:
Refactored config to reduce repetitive code.

hellofromtonya committed:
Removed imports for WP functions.

rgadon107 commented:
## Lessons Learned From This PR

**1) Don’t add a `tearDown()` method to the test class unless something unique needs to be done after each test.** Otherwise, allow the `tearDown()` method from the parent class to reset each test method. 

From the PHP Manual (see ‘Object Inheritance’; https://www.php.net/manual/en/language.oop5.inheritance.php ):

>For example, when you extend a class, the subclass inherits all of the public and protected methods from the parent class. Unless a class overrides those methods, they will retain their original functionality. 

**2) When adding data to the database, validate that the expected data values were stored.** 

**3) When deleting post meta from the database, it’s sufficient to name the post meta key to be deleted.** There’s no need to name the post meta value. 

**4) When deleting data from the database, validate that the expected data values were removed.** 

**5) When working in the context of a test class, use a public method to register to a filter event.** 

There are three types of callables that can be registered to a filter or action; functions, methods, or closures (anonymous or unnamed functions). 

When working in the context of OOP, register _a public method_ to the filter or action. That allows access to the object (including its properties and methods) and the variable `$this` inside the method. 

Avoid registration of an _embedded function_ inside a test method (e.g. __NAMESPACE__ . ‘\my_custom_callback()’ ). It’s not a method (function) of the object. Therefore, _it does not have access to the object_. Per @hellofromtonya: 

>As I noted elsewhere, I’m not a fan of functions within a function (i.e. embedded functions). Why? Readability. Difficulty understanding scope, ie what is available to the embedded function.

Closures (unnamed functions) can be registered to filters or actions. However, since they do not have a string literal name, they are not located by name in PHP’s functions table. They _are registered in memory_ with a long hex number. Per @hellofromtonya: 

>## How to reference a closure

>You have to get the memory ID/location. There are a couple of ways:

    - Storing the closure in a variable.
    - Discover the memory ID/location by looking inside of `$wp_filter` to find the hex code. Messy.

>Both of these approaches are messy and unnecessary. If you need to reference the callback, then use a named function instead.

**6) In object oriented programming, PHP uses an array syntax to register a method to an object when a callable is registered to a filter (or action).** 

### Compare the following: 
#### Procedural programming 

add_filter( ‘my_filter_event’, __NAMESPACE__ . ‘\my_custom_callback’, 10, 3 );

#### Object oriented programming

// Register the callable as a method within the object. 
add_filter( ‘my_filter_event’, [ $this, ‘my_custom_callback` ], 10, 3 ). 

**7) Avoid the use of closures when testing `remove_filter()`, `has_filter()`, `remove_action()`, or `has_action()`.** 

Per @hellofromtonya: 

>## When to use a closure as a callback

>As a general rule, you can use a closure as a callback when that callback is:

    a) _Not reusable_ (not used anywhere else in the code, i.e. a true callback that is bound to _one singular hook_)
    b) Will not be directly unhooked/unregistered, i.e. meaning it remains registered/hooked.

Therefore, in general, we say that a closure callback is not usable within remove_filter(), has_filter(), remove_action(), or has_action().

**8) When unregistering a callback to a filter in order to clean up a test, there’s no need to test that `remove_filter()` is true. Just run the function on the filter.** 

**9) The number of accepted arguments passed to a registered callback must be reflected in the function, method, or closure declaration.** If the number of accepted args is _not_ properly reflected, WP will error. 

**10) `apply_filters() allows one to filter the first parameter passed to the filter.** [ The first parameter is the second argument in `apply_filters()`. ]  Two or more parameters can be passed to the filter, but only the first parameter is filtered. 

**11) PHP’s `func_get_args()` will provide an array list of _all_ the accepted arguments declared by the callback when registered to a filter event.** `func_get_args()` can only be declared inside a function, method, or closure. In the context of a test class, use `func_get_args()` inside a test method. Then the object variable `$this` is accessible within the method. 

**12) After invoking either `func_get_args()` or `func_num_args()`, test the assertion that the expected number of arguments were passed to the method, and they are the expected arguments.** 

**13) Select test method names that accurately reflect what is being evaluated within the method.**

Panukulf reviewed (commented):
Approve

Panukulf commented:
Update

Panukulf commented:
Submit
